File: typescript/years/2015/day1.ts
// https://adventofcode.com/2015/day/1

import once from 'lodash/once';
import { downloadInput } from '../../lib/downloadInput';
import { main } from '../../lib/utils';

const input = once(() => downloadInput('2015', '1'));

main(module, async () => {
  // First Star:
  let floor = 0;
  let firstBasement = null;
  let charNum = 0;
  for (const char of await input()) {
    charNum++;
    if (char === '(') floor++;
    else floor--;
    if (floor === -1 && firstBasement === null) firstBasement = charNum;
  }
  console.log({ floor });

  // Second Star:
  console.log({ firstBasement });
});

File: nim/src/2015/day1.nim
from "../../lib/utils" import fileLines

proc solve(): (int, int) =
  var
    floor = 0
    firstBasement = -1
    charNum = 0

  for line in fileLines("../tmp/2015-1-input.txt"):
    for char in line:
      charNum += 1
      if char == '(': floor += 1
      elif char == ')': floor -= 1
      if floor == -1 and firstBasement == -1:
        firstBasement = charNum

  return (floor, firstBasement)

when isMainModule:
  echo solve()
